### PR TITLE
Normalise models BCI

### PR DESCRIPTION
**Pull request by PierreGtch**
cf #457

### CODE REVIEW

File: braindecode/models/base.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+class EEGModuleMixin:
+    """
+    Mixin class for EEG models.
+
+    Parameters
+    ----------
+    n_channels: int
+        Number of EEG channels.
+    channel_names: list of str
```
Comment (agramfort) [inline braindecode/models/base.py:12]: I would use ch_names to be consistent with mne and some other places in braindecode

File: braindecode/models/base.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    AttributeError: If some input signal-related parameters are not specified and can not be inferred.
+    """
+
+    def __init__(
+            self,
+            n_channels: int = None,
+            channel_names: List[str] = None,
+            input_window_samples: int = None,
+            input_window_seconds: float = None,
+            sfreq: float = None,
```
Comment (agramfort) [inline braindecode/models/base.py:37]:
```suggestion
            n_channels: Optional[int] = None,
            channel_names: Optional[List[str]] = None,
            input_window_samples: Optional[int] = None,
            input_window_seconds: Optional[float] = None,
            sfreq: Optional[float] = None,
```

File: braindecode/models/base.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    Raises
+    ------
+    AttributeError: If some input signal-related parameters are not specified and can not be inferred.
+    """
+
+    def __init__(
+            self,
+            n_channels: int = None,
+            channel_names: List[str] = None,
+            input_window_samples: int = None,
```
Comment (agramfort) [inline braindecode/models/base.py:35]: while we're at it I would rename it to n_times also to be consistent with mne but it's a matter of taste 

cc @robintibor

File: test/unit_tests/models/test_base.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+            sfreq,
+    ):
+        with pytest.raises(ValueError):
+            _ = EEGModuleMixin(
+                n_channels=n_channels,
+                channel_names=channel_names,
+                input_window_samples=input_window_samples,
+                input_window_seconds=input_window_seconds,
+                sfreq=sfreq,
+            )
```
Comment (agramfort) [inline test/unit_tests/models/test_base.py:98]:
```suggestion
@pytest.mark.parametrize(
    'n_channels, channel_names, input_window_samples, input_window_seconds, sfreq',
    [
        (None, None, 200, 2., 100.),
        (1, None, 200, 2., 100.),
        (1, ['ch1'], None, None, None),
        (1, ['ch1'], None, None, 100.),
        (1, ['ch1'], None, 2., None),
        (1, ['ch1'], 200, None, None),
    ]
)
def test_missing_params(
        self,
        n_channels,
        channel_names,
        input_window_samples,
        input_window_seconds,
        sfreq,
):
    module = EEGModuleMixin(
        n_channels=n_channels,
        channel_names=channel_names,
        input_window_samples=input_window_samples,
        input_window_seconds=input_window_seconds,
        sfreq=sfreq,
    )
    with pytest.raises(AttributeError):
        assert module.n_channels == 1
        assert module.channel_names == ['ch1']
        assert module.input_window_samples == 200
        assert module.input_window_seconds == 2.
        assert module.sfreq == 100.

@pytest.mark.parametrize(
    'n_channels, channel_names, input_window_samples, input_window_seconds, sfreq',
    [
        (1, ['ch1'], 200, 2., 100.),
        (None, ['ch1'], 200, 2., 100.),
        (None, ['ch1'], None, 2., 100.),
        (None, ['ch1'], 200, None, 100.),
        (None, ['ch1'], 200, 2., None),
    ]
)
def test_all_params(
        self,
        n_channels,
        channel_names,
        input_window_samples,
        input_window_seconds,
        sfreq,
):
    module = EEGModuleMixin(
        n_channels=n_channels,
        channel_names=channel_names,
        input_window_samples=input_window_samples,
        input_window_seconds=input_window_seconds,
        sfreq=sfreq,
    )
    assert module.n_channels == 1
    assert module.channel_names == ['ch1']
    assert module.input_window_samples == 200
    assert module.input_window_seconds == 2.
    assert module.sfreq == 100.

@pytest.mark.parametrize(
    'n_channels, channel_names, input_window_samples, input_window_seconds, sfreq',
    [
        (2, ['ch1'], 200, 2., 100.),
        (1, ['ch1'], 200, 3., 100.),
    ]
)
def test_incorrect_params(
        self,
        n_channels,
        channel_names,
        input_window_samples,
        input_window_seconds,
        sfreq,
):
    with pytest.raises(ValueError):
        _ = EEGModuleMixin(
            n_channels=n_channels,
            channel_names=channel_names,
            input_window_samples=input_window_samples,
            input_window_seconds=input_window_seconds,
            sfreq=sfreq,
        )
```

File: braindecode/models/base.py [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+#
+# License: BSD-3
+
+from typing import List, Optional
+import warnings
+
+from docstring_inheritance import NumpyDocstringInheritanceMeta
+
+
+def depreciated_args(obj, *args):
```
Comment (robintibor) [inline braindecode/models/base.py:11]:
```suggestion
def deprecated_args(obj, *args):
```

File: braindecode/models/base.py [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+    out_args = []
+    for old_name, new_name, old_val, new_val in args:
+        if old_val is None:
+            out_args.append(new_val)
+            continue
```
Comment (robintibor) [inline braindecode/models/base.py:16]: feel would be nicer if this is an else

File: braindecode/models/base.py [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+    n_chans : int
+        Number of EEG channels.
+    ch_names : list of str
+        Names of the EEG channels.
+    n_times : int
+        Number of time samples of the input window.
+    input_window_seconds : float
+        Length of the input window in seconds.
+    sfreq : float
+        Sampling frequency of the EEG recordings.
```
Comment (cedricrommel) [inline braindecode/models/base.py:46]: Should we say in the doc that args are optional?

**Review by cedricrommel — APPROVED** [review]

**Comment by cedricrommel** [comment]
LGTM. I checked the main EEGModuleMixin class, the deprecation function and 3-4 classes adaptations (not all of them).

**Comment by PierreGtch** [comment]
@robintibor @agramfort @bruAristimunha what do you think about replacing the model param `ch_names` by `chs_info` (a list of dict)? So that we can pass all the information available in `epochs.info[‘chs’]` to the model.

**Pull request merged**